## 1. This is a Jupyter notebook!
<p>A <em>Jupyter notebook</em> is a document that contains text cells (what you're reading right now) and code cells. What is special with a notebook is that it's <em>interactive</em>: You can change or add code cells, and then <em>run</em> a cell by first selecting it and then clicking the <em>run cell</em> button above ( <strong>▶|</strong> Run ) or hitting <code>ctrl + enter</code>. </p>
<p><img src="https://s3.amazonaws.com/assets.datacamp.com/production/project_33/datasets/run_code_cell_image.png" alt=""></p>
<p>The result will be displayed directly in the notebook. You <em>could</em> use a notebook as a simple calculator. For example, it's estimated that on average 256 children were born every minute in 2016. The code cell below calculates how many children were born on average on a day. </p>

In [1]:
# I'm a code cell, click me, then run me!
256 * 60 * 24 # Children × minutes × hours

368640

## 2. Put any code in code cells
<p>But a code cell can contain much more than a simple one-liner! This is a notebook running python and you can put <em>any</em> python code in a code cell (but notebooks can run other languages too, like R). Below is a code cell where we define a whole new function (<code>greet</code>). To show the output of <code>greet</code> we run it last in the code cell as the last value is always printed out. </p>

In [0]:
def greet(first_name, last_name):
    greeting = 'My name is ' + last_name + ', ' + first_name + ' ' + last_name + '!'
    return greeting

# Replace with your first and last name.
# That is, unless your name is already James Bond.
greet('Marc', 'Ishak')

## 3. Jupyter notebooks ♡ data
<p>We've seen that notebooks can display basic objects such as numbers and strings. But notebooks also support the objects used in data science, which makes them great for interactive data analysis!</p>
<p>For example, below we create a <code>pandas</code> DataFrame by reading in a <code>csv</code>-file with the average global temperature for the years 1850 to 2016. If we look at the <code>head</code> of this DataFrame the notebook will render it as a nice-looking table.</p>

In [3]:
# Importing the pandas module
import pandas as pd

# Reading in the global temperature data
global_temp = pd.read_csv('datasets/global_temperature.csv')

# Take a look at the first datapoints
# ... YOUR CODE FOR TASK 3 ...

global_temp

year  degrees_celsius
0    1850             7.74
1    1851             8.09
2    1852             7.97
3    1853             7.93
4    1854             8.19
5    1855             8.12
6    1856             7.90
7    1857             7.71
8    1858             8.13
9    1859             8.20
10   1860             7.78
11   1861             7.81
12   1862             7.49
13   1863             8.15
14   1864             7.94
15   1865             8.13
16   1866             8.19
17   1867             8.28
18   1868             8.09
19   1869             8.32
20   1870             8.08
21   1871             8.05
22   1872             8.12
23   1873             8.24
24   1874             8.38
25   1875             7.87
26   1876             7.99
27   1877             8.49
28   1878             8.78
29   1879             8.14
..    ...              ...
137  1987             9.01
138  1988             9.21
139  1989             8.93
140  1990             9.25
141  1991             9.19
142  1992             8.84
143  1993             8.87
144  1994             9.04
145  1995             9.36
146  1996             9.04
147  1997             9.21
148  1998             9.53
149  1999             9.29
150  2000             9.20
151  2001             9.41
152  2002             9.56
153  2003             9.52
154  2004             9.32
155  2005             9.70
156  2006             9.52
157  2007             9.73
158  2008             9.42
159  2009             9.49
160  2010             9.70
161  2011             9.51
162  2012             9.50
163  2013             9.60
164  2014             9.56
165  2015             9.82
166  2016            10.02

[167 rows x 2 columns]

## 4. Jupyter notebooks ♡ plots
<p>Tables are nice but — as the saying goes — <em>"a plot can show a thousand data points"</em>. Notebooks handle plots as well, but it requires a bit of magic. Here <em>magic</em> does not refer to any arcane rituals but to so-called "magic commands" that affect how the Jupyter notebook works. Magic commands start with either <code>%</code> or <code>%%</code> and the command we need to nicely display plots inline is <code>%matplotlib inline</code>. With this <em>magic</em> in place, all plots created in code cells will automatically be displayed inline. </p>
<p>Let's take a look at the global temperature for the last 150 years.</p>

In [0]:
# Setting up inline plotting using jupyter notebook "magic"
%matplotlib inline

import matplotlib.pyplot as plt

# Plotting global temperature in degrees celsius by year
plt.plot(global_temp['year'], global_temp['degrees_celsius'])

# Adding some nice labels 
plt.xlabel('Year') 
plt.ylabel('Temp (C)') 

## 5. Jupyter notebooks ♡ a lot more
<p>Tables and plots are the most common outputs when doing data analysis, but Jupyter notebooks can render many more types of outputs such as sound, animation, video, etc. Yes, almost anything that can be shown in a modern web browser. This also makes it possible to include <em>interactive widgets</em> directly in the notebook!</p>
<p>For example, this (slightly complicated) code will create an interactive map showing the locations of the three largest smartphone companies in 2016. You can move and zoom the map, and you can click the markers for more info! </p>

In [0]:
# Making a map using the folium module
import folium
phone_map = folium.Map()

# Top three smart phone companies by market share in 2016
companies = [
    {'loc': [37.4970,  127.0266], 'label': 'Samsung: 20.5%'},
    {'loc': [37.3318, -122.0311], 'label': 'Apple: 14.4%'},
    {'loc': [22.5431,  114.0579], 'label': 'Huawei: 8.9%'}] 

# Adding markers to the map
for company in companies:
    marker = folium.Marker(location=company['loc'], popup=company['label'])
    marker.add_to(phone_map)

# The last object in the cell always gets shown in the notebook
phone_map

## 6. Goodbye for now!
<p>This was just a short introduction to Jupyter notebooks, an open source technology that is increasingly used for data science and analysis. I hope you enjoyed it! :)</p>

In [0]:
# Are you ready to get started with  DataCamp projects?
I_am_ready = False

# Ps. 
# Feel free to try out any other stuff in this notebook. 
# It's all yours!